### DISCUSSION TITLE
How to get user public IP address

### DISCUSSION THREAD

**Discussion by blueowlgreenfish** [tokio-rs/axum#2206] [Q&A]
I checked discussion [ Get the client ip from a Request #968 ](https://github.com/tokio-rs/axum/discussions/968).
I tried an example for `into_make_service_with_connect_info`.

    use axum::{
        extract::ConnectInfo,
        routing::get,
        Router,
    };
    use std::net::SocketAddr;

    let app = Router::new().route("/", get(handler));

    async fn handler(ConnectInfo(addr): ConnectInfo<SocketAddr>) -> String {
        // I changed the example code from addr to addr.ip() to get IP address
        format!("Hello {}", addr.ip())
    }

    axum::Server::bind(&"0.0.0.0:3000".parse().unwrap())
        .serve(
            app.into_make_service_with_connect_info::<SocketAddr>()
        )
        .await
        .expect("server failed");

This returns `Hello 127.0.0.1`.
It only returns local address and not a public IP address.
How can I return a public IP address?

**Comment by davidpdrsn** [comment]
Are you using a reverse proxy? If so you need to check the headers it sets.

**Reply by blueowlgreenfish** [reply]
I'm not using a reverse proxy.
I'm using only axum.

**Reply by blueowlgreenfish** [reply]
I don't know if this helps but using [public-ip](https://github.com/avitex/rust-public-ip) returns a public IP address.
But I want to do this using only axum.

**Reply by davidpdrsn** [reply]
Well if you're making requests locally, to a server running locally, your IP will be `127.0.0.1`

**Reply by blueowlgreenfish** [reply]
hmm...
If I use `public-ip` crate it returns my public IP address and not `127.0.0.1` even though I made the request locally.

**Reply by blueowlgreenfish** [reply]
I tested `into_make_service_with_connect_info` on my live web site.
This time, it returned `Hello 10.0.0.2`, which is not my public IP address.

**Reply by davidpdrsn** [reply]
`ConnectInfo<SocketAddr>` returns the [`remote_addr` of the `AddrStream`](https://docs.rs/hyper/latest/hyper/server/conn/struct.AddrStream.html#method.remote_addr).

[From the docs:](https://docs.rs/axum/latest/axum/struct.Router.html#method.into_make_service_with_connect_info)

> This enables extracting things like the client’s remote address.

It doesn't return your public IP. You'll have to get that with something like https://github.com/avitex/rust-public-ip

**Comment by matthiasg** [comment]
I would like to have the client's remote address AND the targeted address (the one where axum runs on). There is a difference for me whether the caller called 127.0.0.1 or the public  ip. Is that possible ?

**Reply by jplatte** [reply]
Yes, that is possible. It's a bit hidden in the documentation, but you have to implement [`Connected<IncomingStream<'_>>`](https://docs.rs/axum/latest/axum/extract/connect_info/trait.Connected.html) for a custom type that gets both the local and remote address from the `IncomingStream`, then use that with `ConnectInfo` / `into_make_service_with_connect_info`.
